Merge Dev to master

LordOfPolls opened the pull request:
## What type of pull request is this?

<!-- Check whichever applies to your PR -->
- [ ] Non-breaking code change
- [x] Breaking code change
- [ ] Documentation change/addition

## Description
<!-- Clearly and concisely describe what this PR is for, and why you feel it should be merged. -->


## Changes
<!-- - A bullet pointed list outlining the changes you have made -->


## Checklist

<!-- These are actions you **must** have taken, if you haven't, your PR will be rejected -->
- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

artem30801 committed:
refactor: Member.has_roles (#252)

Kigstn committed:
fix: bitrate was a bit too high (#255)

ShardlessBun committed:
fix: Safer way to pop waits in Snake.dispatch() (#253)

Kigstn committed:
fix: KeyError in `place_member_data`  (#254)

artem30801 committed:
fix: behaviour of role add and delete events (#250)

leestarb committed:
feat: Improve nickname editing (see desc.) (#241)

* feat: Improve edit nickname

fix broken `Member.edit_nickname`
add new http method to edit guild user
add `Guild.edit_nickname` alias
implement audit log reason
adjust type hints

* refactor!: Use correct name of `modify_current_member`

docs: Fix nickname argument

LordOfPolls committed:
fix: Incorrect typing for SnakeUser guild_ids

Kigstn committed:
feat: add property to voice channel that returns users currently in voice chat (#256)

* feat: add property to voice channel that returns users currently in voice chat

* style: flake

* refactor: `VoiceStateUpdate.before` now correctly displays the previous number of members

leestarb committed:
fix: Missing MFA required permission (#258)

LordOfPolls committed:
feat: Update smart_cache (#257)

- Renames `get_*` to `fetch_*`
- Adds `get_*` as a non-async alias to access cache
- Adds `delete_*` to delete items from cache

LordOfPolls committed:
merge: branch 'master' into dev

zevaryx committed:
feat: 💥 Add Attachment Option to Application (#265)

LordOfPolls committed:
feat: basic modal support (#186)

* feat: basic modal support

* docs: Document return type for wait_for_modal

* feat: Add responses property to ModalContext

* feat: Add more functionality to `wait_for_modal`

You can now pass an author so each wait for waits for a specific user's response
It's documented now too!

* fix: Add ModealContext to `__all__`

LordOfPolls committed:
docs: Modal type hints

Fixes flake 8 errors:
`dis_snek/client/client.py:695:29: ANN201 Missing return type annotation for public function`
`dis_snek/models/discord/modal.py:60:22: ANN201 Missing return type annotation for public function`

Kigstn committed:
fix: embed parsing discord data

Vioshim committed:
Fixes bug in msg commands (None.startswith) (#271)

* Fixes bug in msg commands (None.startswith)

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

silasary committed:
fix: Use `fetch_channel` in place_voice_state_data

leestarb committed:
docs: Fix inaccurate channel typehint (#275)

Kigstn committed:
docs: add modal and attachment option docs, and refactor some docs (#267)

* docs: add modal and attachment option docs, refactor some outdated information, and update the indexing

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* docs: add another listen variation based on #266

* refactor: update example image

* fix: change missing path

* fix: sending with the wrong context

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

leestarb committed:
feat: Allow using event typehint to define listener (#266)

* feat: Allow using event typehint to define listener

ie:
```
@listen()
async def receive_msg(event: events.MessageCreate): ...
```

* refactor: Correct check for internal events typehints in listener

Allow using events typehint in private methods listener

Commit:
feat: audit log support (#270)

* feat: audit log support

* fix: appropriate typehint and docstring

* fix: actually fix docstrings

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

LordOfPolls committed:
fix: Double extension on assets

LordOfPolls committed:
feat: Role icon support

LordOfPolls committed:
feat 💥: Add non-async get_user_guilds method

Also renames `get_user_guilds` to `fetch_user_guilds` hence breaking

LordOfPolls committed:
fix: Attribute error in mutual guilds

LordOfPolls committed:
fix: nested object errors in embed generation

Closes #244

LordOfPolls committed:
Fix: Incorrect casing for TextStyles constant

leestarb committed:
feat: Allow import all triggers from `ext.tasks` (#278)

ie:
`from dis_snek.ext.tasks import IntervalTrigger`

LordOfPolls committed:
feat 💥: Move tasks out of ext namespace

Kigstn committed:
fix: VoiceStateUpdate was not updating the cached channel obj after a member left (#277)

* docs: add modal and attachment option docs, refactor some outdated information, and update the indexing

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* docs: add another listen variation based on #266

* refactor: update example image

* fix: change missing path

* fix: sending with the wrong context

* fix: voice state update removing members correctly

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Kigstn committed:
feat: add ability to get mutual guilds / member instances for users (#272)

* feat: add User.member_instances property

* fix: make BaseUser.mutual_guilds work

* refactor: update docstrings

* fix: didn't want to change that

LordOfPolls committed:
docs: Update docs to account for tasks move

leestarb committed:
perf: Remove unnecessary cache requests (#279)

* perf: Remove unnecessary cache requests

* style: Black format

LordOfPolls committed:
fix: Referenced messages not having a guild

This blindly assumes that references messages are from the current guild if they dont have a guild_id associated. There is no performant way to validate this, and to my knowledge if the guild_id isnt the same as the parent message discord provides a guild id key

LordOfPolls committed:
fix: Guild not being set in channel get messages

Another case of discord not sending expected data

Fixes #276

Commit:
fix: add AuditLogEventType to __all__ (#283)

LordOfPolls committed:
docs: Fix incorrect docstring for client.wait_for_component

LordOfPolls committed:
fix: Incorrect endpoints for sticker fetching

LordOfPolls committed:
fix: Allow sending just stickers

LordOfPolls committed:
fix: Fix dm message references

LordOfPolls committed:
perf: Dont use properties to access user attribute of client

Fixes item 2 of #259

LordOfPolls committed:
feat: Delete guild from cache in `_on_raw_guild_delete`

Fixes item 3 in #259

LordOfPolls committed:
fix: use int guild_id in `_on_raw_guild_delete`

Fixes item 5 in #259

LordOfPolls committed:
refactor: Break up GuildEmojisUpdate call for readability

Fixes item 7 in #259

benwoo1110 committed:
feat: Improve `Webhook.send` handling (see desc.)

* Fix issue with sending file with webhooks
* Add support to send webhooks to thread

leestarb committed:
style: Inline `Asset.animated` (#287)

Also fixes incorrect return type

benwoo1110 committed:
fix: Remove weird and useless imports in webhooks.py

benwoo1110 committed:
docs: Update readme links. (#291)

LordOfPolls committed:
fix: Subcommands not sharing their parents scopes

LordOfPolls committed:
fix: bot.login no longer starts the gateway

leestarb committed:
feat: Add `ThreadChannel.archive` helper method (#292)

zevaryx committed:
refactor 💥: Make all gets non-async, all fetches async (#289)

* refactor: Make all gets non-async, all fetches async

* docs: Fix missing tabs

* refactor: Remove unnecessary try-except in Client.get_*

* fix: Remove guild_id from Guild.get_role and Guild.get_custom_emoji

* refactor: Formatting

* fix: Extra check on emoji to make sure not None

* fix: Snowflake conversion

benwoo1110 committed:
feat: Delete thread from cache in `_on_raw_thread_delete` (#286)

* feat: Delete thread from cache in `_on_raw_thread_delete`

Fixes item 9 in #259

* fix: Missing conversion to snowflake in smart cache.

* fix: Optional conversion for channel's guild_id.

silasary committed:
test: Add tests (#281)

* docs: Start using Bluenix's TypedDicts to document cache inputs

* test: Add pytest to the CI, and a few basic unit tests

* ci: Update ReadTheDocs configuration to get latest python

LordOfPolls committed:
fix: prevent unnecessary object generation in paginators

specifically fixes the select menu bug from #288

leestarb committed:
style: Remove whitespace (#293)

leestarb committed:
feat💥: Improve channel permissions (see desc.) (#284)

* fix!: Refactor channel permission overwrites

Fixes never updated `Channel.permission_overwrites`
- Implement `DictSerializationMixin` in `PermissionOverwrite`
- Rewrite `Channel.permission_overwrites` as list
- Make `Channel.permission_overwrites` public
- Add docs for `id` attribute

* docs: Refactor channel permissions related docs

- Documentate `Channel.permission_overwrites`
- Use docs for attr attributes in PermissionOverwrite
- Add note for `id` attribute in PermissionOverwrite

* fix!: Use `permission_overwrites` instead of `_original_permission_overwrites` in `GuildChannel.clone`

- Remove useless `_original_permission_overwrites`
- Use `map`, `to_dict` for permissions in create channel methods
- Also use correct check for a empty list

* docs: Fix incorrect comment position

* docs: Correct grammar

* fix: KeyError in `permission_overwrites`

* feat: Add `Channel.permissions_for` helper method

Add a note about it in `Member.channel_permissions`

* perf: Refactor `Member.channel_permissions`

Use `filter` and remove unnecessary variables for roles

* refactor: Use `Guild.get_` in `GuildChannel.permissions_for`

* refactor: Remove unnecessary lambdas for permission_overwrites

Commit:
feat: Add suppress embeds method to messages (#294)

* feat: suppress embeds function

* fix: Can return Message object, not None

silasary committed:
ci: Disable codecov comments (#295)

LordOfPolls committed:
feat: Add unicode-emoji support to roles

LordOfPolls committed:
feat: Add support for locale on interactions

LordOfPolls committed:
feat 💥: sync_commands is now default True

LordOfPolls committed:
merge: branch 'master' into dev

zevaryx reviewed (approved):
Everything looks good to me. Have tested most of these with my general-purpose bot and have not had any issues

benwoo1110 committed:
fix: Various channel related type hinting. (#301)

benwoo1110 committed:
fix: Properly handle url passed into embed image. (#299)

benwoo1110 committed:
fix: Use correct fetch methods in `fetch_all_threads` (#300)

benwoo1110 committed:
feat: Add getting of `GuildCategory.news_channels`. (#302)

benwoo1110 committed:
fix: Forgot to pass in username and avatar_url. (#311)

benwoo1110 committed:
fix: Remove unnecessary returns for `remove_reaction` method. (#307)

* fix: Remove unnecessary returns for `remove_reaction` method.

* fix: Missing else clause.

silasary reviewed (approved):
LGTM

AstreaTSS committed:
feat: Add `reply` method to `MessageContext` (#308)

Kigstn reviewed (approved):
Looks good, I'd however delay this until the current breaking prs get merged

AlbertUnruh committed:
feat: Improve debug_exec.py (#315)

* Update debug_exec.py

- if `exec` returns a `Paginator` the Paginator will be send
- the token will *always* be replaced if a str was returned

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

leestarb committed:
feat💥: Improve role icons (see desc.) (#297)

- Add `Role.icon` helper to get role's icon
- Rewrite `icon` and `unicode_emoji` as protected
- Fix typehint of `_unicode_emoji`
- `_unicode_emoji` is now `PartialEmoji` type

artem30801 committed:
Feat: debug scale visual improvements (#304)

* feat: change uptime display in /debug info

* fix: fixed github link in debug scale embed

* feat: display /debug chache as text table

* doc: added more docstrings/typings to table-making function

AstreaTSS committed:
feat💥: allow for iterables of str for msg prefixes (#306)

* feat!: allow for iterables of str for msg prefixes

* fix: fix typehintings for prefixes

* fix: make sync_interactions true to make merging easier

* fix: replace an erroneous return with continue

* refactor: change one-item prefix from set to tuple

benwoo1110 committed:
feat: Improve todict and its usage (#309)

* fix: Do not converter if already a dict instance.

* perf: Use map lambda for to_dict list conversion.

benwoo1110 committed:
feat: Add support for editing webhook. (#312)

AstreaTSS committed:
fix: Mention prefix erroring out (#318)

Commit:
revert: "feat: Improve todict and its usage (#309)" (#320)

This reverts commit 840b3f42247c14e3d267dfef60ba864294fc293a.

benwoo1110 committed:
feat💥: Improve and fix channel edit handling (see desc.) (#303)

* 💥Rename `default_auto_archive_duration`
* Implement processing of various payload data
* Add kwargs for future proving
* Fix type hinting of all edit methods
* Actually tested this time ;p

benwoo1110 committed:
feat: Add support for sending multiple files. (#310)

* feat: Add support for sending multiple files.

* fix: files parameter type hinting.

* fix: Checking for files present.

artem30801 committed:
feat: text format helpers (#314)

* feat: added text formatting utils

* refactor: renames in formatting

* fix: quote_line formatting

* Added proper __all__ and imports, more renames

Commit:
feat: Add custom emoji support to Paginator (#319)

* feat: Customizable button emojis in Paginator

* feat: Default button color for Paginator

LordOfPolls committed:
fix: msg.reference_message_id was an object not an ID

benwoo1110 committed:
fix: Inconsistences with image data processing (#313)

* fix: Inconsistences with image data processing.

* chore: Black format the code

* feat: Use open_file method to open File object.

* fix: Remove old converter import.

AlbertUnruh committed:
feat: Add support for list of paginators to debug exec (#316)

* Update debug_exec.py

- if `exec` returns an iterable containing `Embed`s a Paginator will be created from them

* fix typo

rename `l_results` to `l_result`

* remove unnecessary code (debug_exec.py)

remove `del`

zevaryx committed:
feat: Add `GuildChannel.add_permission` helper method (#321)

* feat: Add `GuildChannel.add_permission` helper method

* fix: Restore erroneously removed import during merge

Co-authored-by: LordOfPolls <22540825+LordOfPolls@users.noreply.github.com>
Co-authored-by: LordOfPolls <ddavidallen13@gmail.com>

benwoo1110 committed:
refactor: Move processing of color from to_dict to a separate method. (#322)

* refactor: Move processing of color from to_dict to a separate method.

* fix: Missing dict type hinting for color.

* fix: Optional type hinting for process_color.

* fix: Remove old converter import.

Co-authored-by: LordOfPolls <22540825+LordOfPolls@users.noreply.github.com>

zevaryx commented on dis_snek/client/utils/formatting.py:
```diff
@@ -0,0 +1,138 @@
+    return f"[{text_style}m"
+
+
+ansi_escape = "[0m"
```
Is there any reason to not have these as `"\u001b"` instead of pasting the character? It's a bit more readable in my opinion, versus just a square. If there's a specific reason, just ignore this

silasary replied:
`\u001b` would be preferable, but it's not important enough to be blocking, I think

zevaryx replied:
I'll resolve this for now then, we'll change it later

silasary reviewed (approved):
Can we please get this merged before more changes land?

leestarb reviewed (approved):
Good

LordOfPolls committed:
docs: Expose modal documentation
